Feature/team badges

karunarapolu opened the pull request:
Fixes issue #2575 
**Design Achievement Badges for Team Overview - Award and Display Team & User Milestones**

**Overview**

This PR implements a comprehensive achievement badge system for teams on the Team Overview Dashboard. Teams and individual team members can now earn badges for milestones such as "Top Contributor," "Early Bird," and other achievement categories. Badges are displayed prominently on the Team Overview Dashboard and user profiles.

##  Features Implemented

### 1. **Team Badge Model & Data Structure**
- **New Model:** `TeamBadge` - Associates badges with teams and optionally with individual users within teams
- **Scope Support:** Badges can be scoped to:
  - `team` - Team-level achievements
  - `topuser_team` - Individual user achievements within a team context
- **Award Types:**
  - **Automatic:** Awarded based on predefined criteria (contributions, closed issues, rankings)
  - **Manual:** Awarded by team admins/leaders
- **Metadata:** Stores reason for award, timestamp, and who awarded it

### 2. **Badge Award System**
Implemented in `feed_signals.py`

### 3. **Admin Interface & Management**
- **Team Badge Admin:** Full CRUD operations for badge management

### 4. **UI/UX Components**

#### **Team Overview Dashboard (`team_overview.html`)**
  - Team-level achievements with icon and title
  - Team member badges under their profile pictures

#### **User Profile**
- Badge showcase section displaying:
  - Team-specific achievements highlighted

### 5. **Automatic Badge Evaluation**
Implemented via `evaluate_team_badges()` and `evaluate_user_team_badges()`:
- Runs periodically via Django signals (`post_save` on Contribution model)
- Evaluates all automatic badges against team metrics

## 6. Testing

### **Test Coverage**

Comprehensive test suite in `website/tests/test_team_badges.py`:

### 7. Json file of the badges in website>fixtures>team_badges.json

### 8. Icons for badges in website>static>img>badges>team_badges

### Badges should be manually added in production database inside Badges model for the team badges to work!!

<img width="1911" height="866" alt="Screenshot 2025-12-28 144124" src="https://github.com/user-attachments/assets/60bcdba9-0a6c-47a8-97f1-383ecb4279c0" />
<img width="1909" height="861" alt="Screenshot 2025-12-28 150352" src="https://github.com/user-attachments/assets/534b4985-02a9-4103-8e87-741ddc484ab8" />
<img width="1909" height="850" alt="Screenshot 2025-12-28 150502" src="https://github.com/user-attachments/assets/fc3aaf5b-eea3-45dd-8bc6-27b875415480" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added team badges to recognize group achievements and contributions.
  * Enabled admins to assign and manage team member badges.
  * Separated user and team badge sections in navigation and profiles for better organization.

* **Bug Fixes**
  * Updated badge display logic to handle zero-user scenarios properly.

* **Documentation**
  * Added admin interface for badge management and tracking.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

karunarapolu committed:
Team Badges feature implemented

karunarapolu committed:
Team Badges feature implemented

karunarapolu committed:
Add tests for team badges

karunarapolu committed:
Ignore .deb files

karunarapolu committed:
Added icons png  and badges json

karunarapolu commented:
@coderabbitai generate docstrings

karunarapolu committed:
Merge branch 'main' into feature/team-badges

karunarapolu committed:
Merge branch 'OWASP-BLT:main' into feature/team-badges

karunarapolu committed:
Merge branch 'OWASP-BLT:main' into feature/team-badges

karunarapolu committed:
Merge branch 'main' into feature/team-badges

karunarapolu committed:
Fixed coderabbit reviews

mdkaifansari04 commented:
@karunarapolu look into coderabbit suggestion, and resolve all the comments by copilot and coderabbit

karunarapolu committed:
Fix coderabbit review issues and combine migrations into one

karunarapolu committed:
Fixed failing tests

Nachiket-Roy commented:
> @coderabbitai look for files that are irrelevant for this PR and should be removed from the PR

@coderabbitai

karunarapolu commented:
@sidd190 bump the migration? 
Also, for the migration 0168, it imported the Badges model which caused the below migrations to fail when I added the column scope to Badge, so i changed a line to fix it. 
In the test_webhook_github.py, for some reason I couldn't pass the tests unless I used os.environ directly rather than settings.GITHUB_WEBHOOK_TOKEN, I can't figure out any other reason or fix.

karunarapolu committed:
Cleaned up code

Nachiket-Roy commented:
@coderabbitai Is the CI error related to how the runtime handles environment variables?

karunarapolu commented:
@Nachiket-Roy `python manage.py makemigrations` worked locally but when creating from scratch, the migrations are failing. In migration 0168, the Badge model is directly being imported from website.models which is causing failures since Badge model has a new column. Since migrations should be isolated from future model changes, importing the model causes Django to try accessing fields that don’t yet exist at that migration stage. 
Replacing the import with `Badge = apps.get_model("website", "Badge")`
fixes this and prevents database setup failures when running migrations from scratch.
I know I should avoid modifying migration files but I think this is an historical bug which needs to be fixed.

Nachiket-Roy commented:
> @Nachiket-Roy `python manage.py makemigrations` worked locally but when creating from scratch, the migrations are failing. In migration 0168, the Badge model is directly being imported from website.models which is causing failures since Badge model has a new column. Since migrations should be isolated from future model changes, importing the model causes Django to try accessing fields that don’t yet exist at that migration stage. Replacing the import with `Badge = apps.get_model("website", "Badge")` fixes this and prevents database setup failures when running migrations from scratch. I know I should avoid modifying migration files but I think this is an historical bug which needs to be fixed.

@coderabbitai could you verify this bug in the main branch and suggest a fix?

Nachiket-Roy commented:
> ## Recommendation
> Could you share the **actual CI error message/logs**? That would help pinpoint whether:
> 
> * It's truly related to environment variable handling
> * It's a different issue altogether
> * There are uncommitted changes causing confusion
> 
> The code as shown in this PR uses proper Django testing patterns and should work correctly in CI.

@coderabbitai "I’m seeing a CI-only failure in GitHub webhook tests.The test test_webhook_assigns_badge_on_first_issue_close returns 503 !=200 in CI, while all other webhook tests pass.Locally, all tests pass once GITHUB_WEBHOOK_SECRET is set, but in CI only this single test hits the 503 path.os.environ is only used in settings.py, and the secret is read dynamically (not at import time). Tests already use override_settings.I tried changing all the settings.GITHUB_WEBHOOK_SECRET to os.environ(GITHUB_WEBHOOK_SECRET)  and the tests passed in CI but I don't think I should change it in the tests and push."

DonnieBLT commented:
Too many files change. Please try again.
